### PR TITLE
ARROW-18016: [CI] Add sccache to r jobs

### PR DESCRIPTION
**Pull request by assignUser**
This is #14390 with reduced scope to make it possible to review. I have added  sccache to all jobs in the `r` group except the conda feedstock jobs: `conda-*-r*` as those pull external images where sccache can't be easily added. I also did not add sccache to the rchk and sanitizer jobs as to not interfere with the results.

### CODE REVIEW

**Comment by assignUser** [comment]
@github-actions crossbow submit test*-r-* homebrew-r-* r-binary-packages

This is `-g r` without the conda jobs

**Comment by assignUser** [comment]
- [ ] homebrew-r-autobrew 	
- [ ] homebrew-r-brew 	
- [ ] r-binary-packages 	
- [ ] test-fedora-r-clang-sanitizer 	
- [ ] test-r-arrow-backwards-compatibility 	
- [ ] test-r-depsource-bundled 	
- [ ] test-r-depsource-system 	
- [ ] test-r-dev-duckdb 	
- [ ] test-r-devdocs 	
- [ ] test-r-gcc-11 	
- [ ] test-r-gcc-12 	
- [ ] test-r-install-local 	
- [ ] test-r-install-local-minsizerel 	
- [ ] test-r-linux-as-cran 	
- [ ] test-r-linux-rchk 	
- [ ] test-r-linux-valgrind 	
- [ ] test-r-minimal-build 	
- [ ] test-r-offline-maximal 	
- [ ] test-r-offline-minimal 	
- [ ] test-r-rhub-debian-gcc-devel-lto-latest 	
- [ ] test-r-rhub-debian-gcc-release-custom-ccache 	
- [ ] test-r-rhub-ubuntu-gcc-release-latest 	
- [ ] test-r-rocker-r-base-latest 	
- [ ] test-r-rstudio-r-base-4.1-opensuse153 	
- [ ] test-r-rstudio-r-base-4.2-centos7-devtoolset-8 	
- [ ] test-r-rstudio-r-base-4.2-focal 	
- [ ] test-r-ubuntu-22.04 	
- [ ] test-r-versions 	
- [ ] test-ubuntu-18.04-r-sanitizer

**Comment by assignUser** [comment]
@github-actions crossbow submit test-r-depsource-bundled test-r-arrow-backwards-compatibility test-r-install* test-r-offline* test-r-minimal-build test-r-rocker* test-r-rhub* test-r-rstudio*

**Comment by assignUser** [comment]
- [ ] homebrew-r-autobrew 	
- [ ] homebrew-r-brew 	
- [x] r-binary-packages 	
- [ ] ~~test-fedora-r-clang-sanitizer~~
- [x] test-r-arrow-backwards-compatibility 	- [working](https://github.com/ursacomputing/crossbow/actions/runs/3380773801/jobs/5613938289#step:7:1749)
- [x] test-r-depsource-bundled - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38916&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=1520)
- [x] test-r-depsource-system - [working](https://github.com/ursacomputing/crossbow/actions/runs/3378518006/jobs/5608823363#step:6:79)
- [x] test-r-dev-duckdb - [working](https://github.com/ursacomputing/crossbow/actions/runs/3378524597/jobs/5608839217#step:6:80)	
- [ ] ~~test-r-devdocs~~ 	
- [x] test-r-gcc-11 - [wroking](https://github.com/ursacomputing/crossbow/actions/runs/3378524597/jobs/5608839217#step:6:80)
- [x] test-r-gcc-12 - [working](https://github.com/ursacomputing/crossbow/actions/runs/3378520328/jobs/5608829085#step:6:79)	
- [ ] test-r-install-local 	- fix added
- [ ] test-r-install-local-minsizerel - fix added
- [ ] test-r-linux-as-cran - unclear why it is not working, due to "as-cran" settings no c++ build logs -> follow-up
- [ ] ~~test-r-linux-rchk~~
- [ ] ~~test-r-linux-valgrind~~ 	
- [x] test-r-minimal-build - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38915&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=77)
- [ ] test-r-offline-maximal - 
- [ ] test-r-offline-minimal - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38917&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=76)
- [ ] test-r-rhub-debian-gcc-devel-lto-latest - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38909&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=1427)
- [ ] test-r-rhub-debian-gcc-release-custom-ccache - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38911&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=1740)
- [ ] test-r-rhub-ubuntu-gcc-release-latest - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38908&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=1762) failure is unrelated test error
- [ ] test-r-rocker-r-base-latest - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38914&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=77)
- [ ] test-r-rstudio-r-base-4.1-opensuse153 - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38910&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=77)	
- [ ] test-r-rstudio-r-base-4.2-centos7-devtoolset-8 - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38912&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=80)
- [ ] test-r-rstudio-r-base-4.2-focal - [working](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38912&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=80)
- [x] test-r-ubuntu-22.04 - [working](https://github.com/ursacomputing/crossbow/actions/runs/3378514340/jobs/5608814653#step:6:79) 	
- [x] test-r-versions - [working](https://github.com/ursacomputing/crossbow/actions/runs/3378524036/jobs/5608838294#step:6:80) 	
- [ ] ~~test-ubuntu-18.04-r-sanitizer~~

**Comment by assignUser** [comment]
@github-actions crossbow submit test-r-offline-maximal test-r-install-local*

**Comment by assignUser** [comment]
@kou @raulcd this should be ready for review. I will investigate the homebrew builds and the linux-as-cran builds in a followup because I don't see any obvious issues and it should work imo.

The three builds I just started should work now with the fixes I pushed. I linked the logs for everything else that shows sccache is being used.

File: ci/scripts/install_sccache.sh [hunk 1]
```diff
@@ -41,14 +41,22 @@ SCCACHE_ARCHIVE=sccache.tar.gz
...
+fi
+
+sha256sum $SHA_ARGS $SCCACHE_ARCHIVE.sha256
 
 if [ ! -d $PREFIX ]; then
     mkdir -p $PREFIX
 fi
 
-tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory $PREFIX --wildcards sccache*/sccache* 
+tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory $PREFIX --exclude="sccache*/*E*E*"
```
Comment (kou) [inline ci/scripts/install_sccache.sh:59]: Could you add an expected result with an example as comment?
This ignores `LICENSE` and `README.md` by `*E*E*`?

File: dev/tasks/r/github.linux.cran.yml [hunk 2]
```diff
@@ -44,7 +44,9 @@ jobs:
 
       - name: Docker Run
         shell: bash
-        # setting ARROW_SOURCE_HOME='' here ensures that we use the cpp source copied into tools/
+        env:
+        {{ macros.github_set_sccache_envvars()|indent(8)}}
+          # setting ARROW_SOURCE_HOME='' here ensures that we use the cpp source copied into tools/
```
Comment (kou) [inline dev/tasks/r/github.linux.cran.yml:49]:
```suggestion
        # setting ARROW_SOURCE_HOME='' here ensures that we use the cpp source copied into tools/
```

File: dev/tasks/tasks.yml [hunk 3]
```diff
@@ -157,6 +157,7 @@ groups:
     - example-*
     - wheel-*
     - python-sdist
+    - r-binary-packages
```
Comment (kou) [inline dev/tasks/tasks.yml:160]: Do we need this?
I think that the `nightly` group isn't maintained (but misused: ARROW-18094).

**Review by kou — APPROVED** [review]
+1

File: dev/tasks/tasks.yml (hunk 3 above)
Comment (assignUser) [reply dev/tasks/tasks.yml:160]: I was initially using it for testing but we can drop it. I agree on ARROW-18094

File: ci/scripts/install_sccache.sh (hunk 1 above)
Comment (assignUser) [reply ci/scripts/install_sccache.sh:59]: Yes exactly, will add a comment, the `wildcards` version was more intuitive but does not work on busybox.

File: ci/scripts/install_sccache.sh [hunk 4]
```diff
@@ -41,14 +41,23 @@ SCCACHE_ARCHIVE=sccache.tar.gz
...
+fi
+
+sha256sum $SHA_ARGS $SCCACHE_ARCHIVE.sha256
 
 if [ ! -d $PREFIX ]; then
     mkdir -p $PREFIX
 fi
 
-tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory $PREFIX --wildcards sccache*/sccache* 
+# Extract only the sccache binary into $PREFIX and ignore README and LCIENSE
```
Comment (kou) [inline ci/scripts/install_sccache.sh:59]:
```suggestion
# Extract only the sccache binary into $PREFIX and ignore README and LCIENSE.
# --wildcards doesn't work on busybox.
```

**Pull request merged**